pkg/profile/chart_builder.go

```diff
@@ -0,0 +1,116 @@
+package profile
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta1"
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	profilesv1 "github.com/weaveworks/profiles/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ChartBuilder struct {
+	GitRepositoryNamespace string
+	GitRepositoryName      string
+	RootDir                string
+}
+
+func (c *ChartBuilder) Build(artifact profilesv1.Artifact, installation profilesv1.ProfileInstallation, definition profilesv1.ProfileDefinition) ([]Artifact, error) {
+	a := Artifact{Name: artifact.Name}
+	helmRelease := c.makeHelmRelease(artifact, installation, definition)
+	a.Objects = append(a.Objects, helmRelease)
+	if artifact.Chart.Path != "" {
+		if c.GitRepositoryNamespace == "" && c.GitRepositoryName == "" {
+			return nil, fmt.Errorf("in case of local resources, the flux gitrepository object's details must be provided")
+		}
+		helmRelease.Spec.Chart.Spec.Chart = filepath.Join(c.RootDir, "artifacts", artifact.Name, artifact.Chart.Path)
+		branch := installation.Spec.Source.Branch
+		if installation.Spec.Source.Tag != "" {
+			branch = installation.Spec.Source.Tag
+		}
+		a.RepoURL = installation.Spec.Source.URL
+		a.SparseFolder = definition.Name
+		a.Branch = branch
+		a.PathsToCopy = append(a.PathsToCopy, artifact.Chart.Path)
+	}
+	if artifact.Chart.URL != "" {
+		helmRepository := c.makeHelmRepository(artifact.Chart.URL, artifact.Chart.Name, installation)
+		a.Objects = append(a.Objects, helmRepository)
+	}
+	return []Artifact{a}, nil
+}
+
+func (c *ChartBuilder) makeHelmRelease(artifact profilesv1.Artifact, installation profilesv1.ProfileInstallation, definition profilesv1.ProfileDefinition) *helmv2.HelmRelease {
+	var helmChartSpec helmv2.HelmChartTemplateSpec
+	if artifact.Chart.Path != "" {
+		helmChartSpec = c.makeGitChartSpec(path.Join(installation.Spec.Source.Path, artifact.Chart.Path))
+	} else if artifact.Chart != nil {
+		helmChartSpec = c.makeHelmChartSpec(artifact.Chart.Name, artifact.Chart.Version, installation)
+	}
+	helmRelease := &helmv2.HelmRelease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      makeArtifactName(artifact.Name, installation, definition),
+			Namespace: installation.ObjectMeta.Namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       helmv2.HelmReleaseKind,
+			APIVersion: helmv2.GroupVersion.String(),
+		},
+		Spec: helmv2.HelmReleaseSpec{
+			Chart: helmv2.HelmChartTemplate{
+				Spec: helmChartSpec,
+			},
+			Values:     installation.Spec.Values,
+			ValuesFrom: installation.Spec.ValuesFrom,
+		},
+	}
+	return helmRelease
+}
+
+func (c *ChartBuilder) makeHelmRepository(url string, name string, installation profilesv1.ProfileInstallation) *sourcev1.HelmRepository {
```

Comment:
In some cases, installation is not needed, only the namespace. I'm fixing passing around the installation and definition too much.